### PR TITLE
Introduce new commands

### LINKED ISSUE #31: META: Turning `hippofactory` into `hippo`, the CLI for Hippo
We've been talking about using Hippofactory as the basis for the general Hippo CLI. I think these are the major things we would need for a full-fledged UI (some of which are done, and some of which tie closely to demos):

In all of these cases, the assumption is that APP_NAME and other details can be gotten from `./HIPPOFACTS` and that some information can be gained from contacting the API server (e.g. which channels exist, which is default, etc)

## Required

- [ ] `hippo login URL` - login to a Hippo server, after which all ops go to that server
- [ ] `hippo create|remove [APP_NAME]` - create and remove apps. If app name is not specified, use name in `./HIPPOFACTS`
- [ ] `hippo push [CHANNEL]` - push to Bindle and notify the current app on the given (or default) channel.
- [ ] `hippo logs [CHANNEL]` - view logs for the given (or default) channel.
- [ ] `hippo run` - Run the code in a local Wagi instance (for local testing)
- [ ]  `hippo prepare` - Prepare a local bindle, but don't upload
- [ ]  `hippo env set NAME VALUE [CHANNEL]` - Set an env var
- [ ]  `hippo env get|remove NAME VALUE [CHANNEL]` - Set an env var
- [ ] `hippo domain set DNS_NAME [CHANNEL]` - Set the DNS name for an app 
- [ ] `hippo domain get|remove [CHANNEL]` - Get/remove the DNS name for an app. Remove sets it back to the default
- [ ]  `hippo cert set CERT [CHANNEL]` - Set a TLS/SSL cert for a channel (This might need to split cert and key into two files... we can figure that out)
- [ ]  `hippo cert remove [CHANNEL]` - Remove TLS/SSL cert for a channel

## Optional (Possibly unnecessary)
- [ ] `hippo init` - Create a new `HIPPOFACTS` file

## Optional (Maybe UI only)
- [ ] `hippo rollback [CHANNEL]`
- [ ] `hippo pin|unpin CHANNEL RELEASE_VERSION` - pin a channel to a version
- [ ] `hippo channel create|remove CHANNEL`

## Optional (So user doesn't need bindle client, too)
- [ ] `hippo bindle search STRING` - Search the bindle repo, useful for adding dependencies
- [ ] `hippo bindle yank VERSION` - Mark a particular bindle as yanked

## Conventions:

- The convention for commands is `hippo [NOUN] VERB`, with the exception of `hippo logs` (I am open to a verb that means "to read logs" and has some traction in other CLIs)
  - The most common user actions will be on apps, so instead of doing `hippo app create`, `hippo app logs`, etc., I reduced it to `hippo create` and `hippo logs`. This is consistent with the vast majority of CLIs out there.
  - Subgroups are then used for less frequently accessed commands/objects: `hippo channel create`
- I used `remove` instead of `delete` per the [Helm UX research](https://github.com/helm/community/blob/124ab295bd37a4518d96f8e6c989ec73fc1e357e/hips/archives/helm/helm-v3/research/package-manager-ux.md) on the topic, and we could use the `rm` alias if we want

### PR DESCRIPTION
**Pull request by bacongobbler**
This restructures the library and introduces several new commands:

- hippo app <add|remove>
- hippo auth <register|login|logout|whoami>
- hippo certificate <add|remove>
- hippo channel <add|remove>
- hippo env <add|remove>
- hippo revision add

These commands are now possible thanks to the API updates provided in Hippo 0.2.

The packages received an overhaul as well:

- all files preceded with `bindle_` were moved to module `bindle`
- the CLI is now hosted in module `cli`
- all commands are defined in module `cli::commands`
- the hippo client has moved to module `hippo`

closes #31

Signed-off-by: Matthew Fisher <matt.fisher@fermyon.com>

### CODE REVIEW

**Comment by bacongobbler** [comment]
NOTE: only `hippo auth register` is partially implemented at the moment, but the majority of the skeleton framework has been laid out. Now I have to go back and implement the commands.

File: src/cli/commands/mod.rs
Comment (radu-matei) [inline src/cli/commands/mod.rs:18]: nit: would it be more readable if command enums were directly named like `AppCommands` or `AuthCommands`?

File: src/cli/commands/auth.rs
Comment (radu-matei) [reply src/cli/commands/auth.rs:14]: Also, when the validation fails without an exception, the CLI will return a success message, even if the account hasn't been register successfully:

```
❯ hippo auth register --username root --password PassWord! --danger-accept-invalid-certs https://localhost:5309
Registered new account: root
```

Server logs:

```
info: Hippo.Application.Accounts.Commands.CreateAccountCommand[0]
      Hippo Request: CreateAccountCommand   Hippo.Application.Accounts.Commands.CreateAccountCommand
warn: Microsoft.AspNetCore.Identity.UserManager[14]
      User password validation failed: PasswordRequiresDigit.
```

File: src/cli/commands/app.rs
Comment (radu-matei) [inline src/cli/commands/app.rs:4]: Not blocking this PR — do we also want a way to list the apps?

File: src/cli/commands/app.rs
Comment (radu-matei) [inline src/cli/commands/app.rs:15]: Creating an application works great — I can validate that in the UI:

```
❯ hippo app add testapp radu/testapp
Added App testapp (ID = '2f0c046f-ad29-4846-892b-b850e7d0eba7')
IMPORTANT: save this App ID for later - you will need it to update and/or delete the App (for now)
```

Deleting the app doesn't seem to work:

```
❯ hippo app remove 2f0c046f-ad29-4846-892b-b850e7d0eba7
Error: error in response: status code 404 Not Found
```

File: src/cli/commands/environment_variable.rs
Comment (radu-matei) [inline src/cli/commands/environment_variable.rs:13]: Creating a channel works, but removing it returns a 404:

```
❯ hippo channel add staging 2f0c046f-ad29-4846-892b-b850e7d0eba7
Added Channel staging (ID = 'e5114347-fba8-4d19-93c0-59a4cd3e54b4')
IMPORTANT: save this Channel ID for later - you will need it to update and/or delete the Channel (for now)
```

```
❯ hippo channel remove e5114347-fba8-4d19-93c0-59a4cd3e54b4
Error: error in response: status code 404 Not Found
```

File: src/cli/commands/channel.rs
Comment (radu-matei) [inline src/cli/commands/channel.rs:22]: Creating a channel works, but removing it returns a 404:

```
❯ hippo channel add staging 2f0c046f-ad29-4846-892b-b850e7d0eba7
Added Channel staging (ID = 'e5114347-fba8-4d19-93c0-59a4cd3e54b4')
IMPORTANT: save this Channel ID for later - you will need it to update and/or delete the Channel (for now)

❯ hippo channel remove e5114347-fba8-4d19-93c0-59a4cd3e54b4
Error: error in response: status code 404 Not Found
```

File: src/cli/commands/auth.rs
Comment (bacongobbler) [reply src/cli/commands/auth.rs:14]: Yeah this is a server-side issue. Hippo should be returning a proper error response. We can fix that up in the API :)

Comment (bacongobbler) [reply src/cli/commands/auth.rs:14]: The latter case is also a server-side issue. Nice catch.

File: src/cli/commands/app.rs
Comment (bacongobbler) [reply src/cli/commands/app.rs:4]: Yes, though there’s no support for that in the API at this time. I also have to expose /edit APIs much like what you can do in the web UI

File: src/cli/commands/mod.rs
Comment (bacongobbler) [reply src/cli/commands/mod.rs:18]: I figured the namespace would be enough of an indicator… Is the convention in Rust to “stutter” and re-repeat the namespace in the type name (`app::AppCommands`)? I thought most language linters discourage that convention

File: src/cli/commands/app.rs
Comment (bacongobbler) [reply src/cli/commands/app.rs:15]: Good catch. I think a 404 is the expected response, so maybe we need to capture that instead of casting via `?`

File: src/cli/commands/mod.rs
Comment (radu-matei) [reply src/cli/commands/mod.rs:18]: You would `use app::AppCommands` in the top level, and throughout the file you'd use `AppCommands` directly.

Comment (bacongobbler) [reply src/cli/commands/mod.rs:18]: Clippy has a linting rule against types that stutter (`clippy::module_name_repetitions`), but it is disabled by default. Names like `app::AppCommands` is generally discouraged.

File: src/cli/commands/auth.rs
Comment (bacongobbler) [reply src/cli/commands/auth.rs:14]: > Should there be some local validation here before sending to the server, as well as on the server?

I don't see a reason why we should validate locally... That just means we'll have to update the same validation rules in two different locations, and that could lead to validation drift. IMO I'd rather put this validation all on the server so every client validates objects the same.

Comment (bacongobbler) [reply src/cli/commands/auth.rs:14]: filed https://github.com/deislabs/hippo/issues/398 and https://github.com/deislabs/hippo/issues/399 - thanks!

File: src/cli/mod.rs
Comment (bacongobbler) [reply src/cli/mod.rs:62]: good call. Done

File: src/cli/commands/app.rs
Comment (bacongobbler) [reply src/cli/commands/app.rs:15]: Hmm... The server should be responding with a 204 no content according to https://github.com/deislabs/hippo/blob/bae252ef73038ec6e473d852c384aef70695e84d/src/Web/Api/AppController.cs#L41

Will look into this further

File: src/cli/commands/channel.rs
Comment (bacongobbler) [reply src/cli/commands/channel.rs:11]: fixed

File: src/cli/commands/app.rs
Comment (bacongobbler) [reply src/cli/commands/app.rs:15]: This is happening in the UI as well, so I'm lead to believe this is a server-side issue. Once I determine the cause I'll try to fix it before merging this

Comment (bacongobbler) [reply src/cli/commands/app.rs:15]: identified the issue in https://github.com/deislabs/hippo/pull/401 - misconfigured API routes

File: src/cli/commands/environment_variable.rs
Comment (bacongobbler) [reply src/cli/commands/environment_variable.rs:15]: https://github.com/deislabs/hippo-cli/pull/99#discussion_r794740312

File: src/cli/commands/channel.rs
Comment (bacongobbler) [reply src/cli/commands/channel.rs:30]: https://github.com/deislabs/hippo-cli/pull/99#discussion_r794740312

**Comment by bacongobbler** [comment]
Happy path workflow:

```
$ hippo auth register --danger-accept-invalid-certs
Enter username: bacongobbler
Enter password: [hidden]
Registered bacongobbler
$ hippo auth login --danger-accept-invalid-certs
Enter Hippo username: bacongobbler
Enter Hippo password: [hidden]
Logged in as bacongobbler
$ hippo app add toast-on-demand toast-on-demand
Added App toast-on-demand (ID = 'fbfc8c40-7e39-4ce7-8693-ee7643f564a0')
IMPORTANT: save this App ID for later - you will need it to update and/or delete the App (for now)
$ hippo channel add latest fbfc8c40-7e39-4ce7-8693-ee7643f564a0
Added Channel latest (ID = '29f28347-9f5d-4323-92dc-5a7172afff8a')
IMPORTANT: save this Channel ID for later - you will need it to update and/or delete the Channel (for now)
$ hippo push . -v production
Pushed toast-on-demand/0.3.0
$ curl -H "Host: latest.toast-on-demand.hippofactory.io" https://localhost:5309 -k
<h1>Welcome to the Toast on Demand web site</h1>
<p>Our mission is to provide you with delicious and nutritious toast - where you want it, when you want it, and as browned as you want it.</p>
<h2>TOAST OF THE DAY</h2>
<p>Today's special is RAW TOAST</p>
<p>Served straight or on the rocks.</p>
<p><img src='assets/images/raw_toast.jpg' /></p>
```

File: src/cli/commands/auth.rs
Comment (bacongobbler) [reply src/cli/commands/auth.rs:15]: New error message:

```
><> hippo auth register --danger-accept-invalid-certs
Enter username: bacongobbler
Enter password: [hidden]
Error: {"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"One or more validation errors occurred.","status":400,"errors":{"Password":["The length of 'Password' must be at least 8 characters. You entered 3 characters."]}}
```

**Review by radu-matei — APPROVED** [review]
I think all the comments were addressed.

LGTM!

**Pull request merged**
